### PR TITLE
Configurable FREEZE pin state

### PR DESCRIPTION
**Pull request by ellensp**
### Description

Add trivial feature to Invert the freeze pin logic as requested in FR

### Requirements

FREEZE_FEATURE

### Benefits

Pin can be logically inverted. 

### Configurations

Added keyword to Configuration_adv.h 
#define INVERT_FREEZE_PIN

### Related Issues
https://github.com/MarlinFirmware/Marlin/issues/23942

### CODE REVIEW

**Comment by GMagician** [comment]
my 2cent:
1) There is in code a pulllup that should be changed to pulldown. 
2) Isn't better to use logic used elsewhere? like FREEZE_ACTIVE HIGH/LOW?

**Comment by thisiskeithb** [comment]
> Isn't better to use logic used elsewhere? like FREEZE_ACTIVE HIGH/LOW?

There seems to be a mix of HIGH/LOW and 0/1 usage in the config, so either of those should work.

**Comment by KillerBug** [comment]
Since it's a safety thing it's a good idea for an open to be the freeze state in case a wire fails.  With a pull-up you can just connect to ground and trigger any time it's not grounded.  Or at least that's my theory.

I have these changes in my firmware and it's pretty sweet having an emergency stop button that allows for resume.  My e-stop switch is actually two normally closed switches.  Using one for marlin with this feature and the other breaks the water-protect-circuit and shuts off the laser power supply directly.

Something similar could be done with a pull-down but I don't think my board (octopus) even supports pull-downs.  At least they don't work with limit switches.

**Comment by thinkyhead** [comment]
Although the request literally asked for an "invert" option, the new approach is to have a "state" option defined as `HIGH` or `LOW` instead.

**Comment by GMagician** [comment]
@KillerBug ok if SoC doesn't support pulldown but when you use open collector signals, if you close to gnd you need a pullup and when you close at vcc  you need a pulldown. Push-pull signal don't need pullup/down. On some board correct pull is done on pcb.Some limit switch (physical switch) have a 3 pin switch and acts like a push-pull signal

**Comment by KillerBug** [comment]
@GMagician - What I mean is I know of at least one board that doesn't support pull-downs but I know of no boards that don't support pull-ups.  Maybe that's just been my luck and there are lots of pull-down only boards?

Maybe it should be able to support pull up or pull down, similar to how limit switches are setup?

**Comment by thinkyhead** [comment]
The current behavior is to init the pin with pull-up to voltage, and the switch is assumed to be NC. If the pin is connected to an NO switch instead, the pull-up is probably not appropriate, but a pull-down may be. In the past I based the choice of pull-up or pull-down on whether the switch is NC or NO, but that is not always ideal either, so for some pins/option we allow the user to configure how the pin should be initialized.

**Comment by KillerBug** [comment]
> The current behavior is to init the pin with pull-up to voltage, and the switch is assumed to be NC.

You are correct that in Marlin 2.0.9.3 the behavior is to pull-up the pin.  However, the switch is assumed to be NO.  The changes made by ellensp keeps the pull up but allows for selecting a NC switch (activate when high).

**Comment by thinkyhead** [comment]
> You are correct that in Marlin 2.0.9.3 the behavior is to pull-up the pin. However, the switch is assumed to be NO.

The code before this change:
```cpp
frozen = !READ(FREEZE_PIN);
```
That's the logic for an NC switch, not for an NO switch.

File: Marlin/src/MarlinCore.cpp [hunk 1]
```diff
@@ -476,7 +476,7 @@ inline void manage_inactivity(const bool no_stepper_sleep=false) {
   #endif
 
   #if HAS_FREEZE_PIN
-    Stepper::frozen = !READ(FREEZE_PIN);
+    Stepper::frozen = READ(FREEZE_PIN) ^ DISABLED(INVERT_FREEZE_PIN);
```
Comment (GMagician) [inline Marlin/src/MarlinCore.cpp:479]: Is INVERT_FREEZE_PIN still present?

Comment (thisiskeithb) [reply Marlin/src/MarlinCore.cpp:479]: Nope. This will need another patch.

Comment (ellensp) [reply Marlin/src/MarlinCore.cpp:479]: The PR is all messed up now.
FREEZE_STATE only enables pullup or pull down and does not invert the logic level.
The entire original purpose of PR has been lost.

To many busy bodies and a rouge main programmer.
It enough to drive me around the bend.

**Comment by KillerBug** [comment]
The pull-up was never the issue in the first place...this all started because the function was written for NO switches and I wanted to use a NC switch.  The original fix by ellensp did that.

> frozen = !READ(FREEZE_PIN);
> That's the logic for an NC switch, not for an NO switch.

The description of the function is "Specified pin has pullup and connecting to ground will instantly pause motion."  This description is correct.  A NC switch would pause motion when not pressed, and allow motion when pressed; the exact opposite of what you would want.  When this was reversed by ellensp it works correctly.

If I am reading this correct, frozen=1 means it is frozen, frozen=0 means it isn't.
The function reads the pin.  It's a pull-up so it will be 1 when not connected and 0 when connected.
With a NC switch to ground it will read 0 when the switch is unpressed, and 1 when it is pressed.
But it will be inverted, so we get 0 when pressed.
So when we press the button that should stop all motion, frozen=0.

**Comment by GMagician** [comment]
For a NO or NC input you should always have a pullup/down depending on what you close to (down if you close to vcc or up if you close to gnd), then you have to define if input freeze with 0V or vcc.
With current solution:
"active state"=low you may only connect NO that closes to gnd
"active state"=high you may only NO that closes to vcc
With original ellensp solution:
"no invert" you may only connect NO that closes to gnd
"invert" you may only connect NC that closes to gnd

On industrial machines logic is always positive so switches give vcc when closed

I think that a complete solution should let you use NO/NC that may close to gnd or vcc, from coding side is just a double xor but maybe configuration is a little more problematic (confusion may arise)

**Pull request merged**